ArjunaJTA/jta/tests/classes/com/hp/mwtests/ts/jta/recovery/RecoveryScanTest.java

```diff
@@ -0,0 +1,167 @@
+        manager.startRecoveryManagerThread(); // start periodic recovery
+    }
+
+    @AfterClass
+    public static void afterClass() throws Exception {
+        manager.terminate();
+        XATestResourceXARecovery.setUseFaultyResources(true);
+    }
+
+    @Test
+    public void testRecoveryMonitorWithFailure() throws Exception {
+        XATestResourceXARecovery.setUseFaultyResources(true);
+        UserTransaction ut = com.arjuna.ats.jta.UserTransaction.userTransaction();
+        ut.begin();
+        TransactionImple tx = TransactionImple.getTransaction();
+
+        // enlist a resource that behaves correctly
+        assertTrue(tx.enlistResource(new XATestResource(XATestResource.OK_JNDI_NAME, false)));
+        // enlist a resource that throws an exception from recover
+        assertTrue(tx.enlistResource(new XATestResource(XATestResource.FAULTY_JNDI_NAME, true)));
+
+        ut.commit();
+
+        manager.scan();
+
+        assertTrue(jtaLogger.isRecoveryProblems());
+    }
+
+    @Test
+    public void testRecoveryMonitorWithSuccess() throws Exception {
+        XAResource dummy = new XAResource() {
+            @Override
+            public void commit(Xid xid, boolean b) throws XAException {
+            }
+
+            @Override
+            public void end(Xid xid, int i) throws XAException {
+            }
+
+            @Override
+            public void forget(Xid xid) throws XAException {
+            }
+
+            @Override
+            public int getTransactionTimeout() throws XAException {
+                return 0;
+            }
+
+            @Override
+            public boolean isSameRM(XAResource xaResource) throws XAException {
+                return false;
+            }
+
+            @Override
+            public int prepare(Xid xid) throws XAException {
+                return 0;
+            }
+
+            @Override
+            public Xid[] recover(int i) throws XAException {
+                return new Xid[0];
+            }
+
+            @Override
+            public void rollback(Xid xid) throws XAException {
+            }
+
+            @Override
+            public boolean setTransactionTimeout(int i) throws XAException {
+                return false;
+            }
+
+            @Override
+            public void start(Xid xid, int i) throws XAException {
+            }
+        };
+
+        XATestResourceXARecovery.setUseFaultyResources(false);
+        UserTransaction ut = com.arjuna.ats.jta.UserTransaction.userTransaction();
+        ut.begin();
+        TransactionImple tx = TransactionImple.getTransaction();
+
+        // enlist a resource that behaves correctly
+        assertTrue(tx.enlistResource(dummy));
+        assertTrue(tx.enlistResource(dummy));
+
+        ut.commit();
+
+        manager.scan();
+
+        assertFalse(jtaLogger.isRecoveryProblems());
+    }
+
+    @Test
+    public void testBoth() throws Exception{
+        testRecoveryMonitorWithFailure();
```

Comment:
Good test, this is the one I think we need to see where a resource transitions from failing to passing